lib/auth/accountrecovery.go

```diff
@@ -103,66 +89,61 @@ func (a *Server) StartAccountRecovery(ctx context.Context, req *proto.StartAccou
 	return token, nil
 }
 
-// verifyCodeWithRecoveryLock counts number of failed attempts at providing a valid recovery code.
-// After MaxAccountRecoveryAttempts, user is temporarily locked from further attempts at recovering and also
-// locked from logging in. Modeled after existing function WithUserLock.
-func (a *Server) verifyCodeWithRecoveryLock(ctx context.Context, username string, recoveryCode []byte) error {
-	user, err := a.Services.GetUser(ctx, username, false)
-	switch {
-	case trace.IsNotFound(err):
-		// If user is not found, still authenticate. It should always return an error.
-		// This prevents username oracles and timing attacks.
-		return a.verifyRecoveryCode(ctx, username, recoveryCode)
-	case err != nil:
+// verifyRecoveryCodeWithRecord validates the recovery code for the user and will unlock their account if
+// the code is valid.  If the code is invalid, a failed recovery attempt will be recorded.
+func (a *Server) verifyRecoveryCodeWithRecord(ctx context.Context, username string, recoveryCode []byte) (errResult error) {
+	_, err := a.Services.GetUser(ctx, username, false)
+	if err != nil && !trace.IsNotFound(err) {
+		// In the case of not found, we still want to perform the comparison.
+		// It will result in an error but this avoids timing attacks which expose account presence.
 		log.Error(trace.DebugReport(err))
 		return trace.AccessDenied(startRecoveryGenericErrMsg)
 	}
+	hasRecoveryCodes := false
+	defer func() { // check for result condition in defer func and send the appropriate audit event
+		event := &apievents.RecoveryCodeUsed{
+			Metadata: apievents.Metadata{
+				Type: events.RecoveryCodeUsedEvent,
+				Code: events.RecoveryCodeUseSuccessCode,
+			},
+			UserMetadata: authz.ClientUserMetadataWithUser(ctx, username),
+			Status: apievents.Status{
+				Success: errResult == nil,
+			},
+		}
+		if errResult == nil {
+			if err := a.emitter.EmitAuditEvent(a.closeCtx, event); err != nil {
+				log.WithFields(logrus.Fields{"user": username}).Warn("Failed to emit account recovery code used event.")
+			}
+		} else {
+			event.Metadata.Code = events.RecoveryCodeUseFailureCode
+			traceErr := trace.NotFound("invalid user or user does not have recovery codes")
+
+			if hasRecoveryCodes {
+				traceErr = trace.BadParameter("recovery code did not match")
+			}
+
+			event.Status.Error = traceErr.Error()
+			event.Status.UserMessage = traceErr.Error()
```

Comment:
Do we need to create an error object only to pull the string out of it?

Creating an error with the trace package is a costly operation since it pulls a full stack trace. If there are recovery codes, we're collecting not one but two stack traces here, only to throw them away.

I would get rid of `traceErr` completely, and just set `event.Status.Error = "recovery code did not match"`.

Reply:
I just moved the code without change.  In the most recent commit I updated this to only set the `event.Status.Error` string directly.